File: 3170/solution.py
from heapq import heappop, heappush


class Solution:
    def clearStars(self, s: str) -> str:
        heap = []
        for i, v in enumerate(s):
            if v == "*":
                heappop(heap)
            else:
                heappush(heap, [v, -i])
        heap.sort(key=lambda x: -x[1])
        return "".join([item[0] for item in heap])
        # time O(nlogn)
        # space O(n)

File: 3170/solution.java
import java.util.*;

class Solution {
    public String clearStars(String s) {
        List<Deque<Integer>> stacks = new ArrayList<>();
        for (int i = 0; i < 26; i++) {
            stacks.add(new ArrayDeque<>());
        }

        char[] chars = s.toCharArray();
        for (int i = 0; i < s.length(); i++) {
            char curr = s.charAt(i);
            if (curr != '*') {
                stacks.get(curr - 'a').add(i);
                continue;
            }
            for (int j = 0; j < 26; j++) {
                if (!stacks.get(j).isEmpty()) {
                    int charIndex = stacks.get(j).pollLast();
                    chars[charIndex] = '*';
                    break;
                }
            }
        }
        StringBuilder res = new StringBuilder();
        for (int i = 0; i < s.length(); i++) {
            if (chars[i] != '*') {
                res.append(s.charAt(i));
            }
        }
        return res.toString();
    }
}

// time O(n * uniqueChars)
// space O(n)